File: lab2/insert.py
import unittest


def insertion_sort(arr):
    n = len(arr)
    for i in range(1, n):
        key = arr[i]
        j = i - 1

        while j >= 0 and arr[j] > key:
            arr[j + 1] = arr[j]
            j -= 1
        arr[j + 1] = key


def load_array(filename):
    with open(filename, "r") as file:
        return list(map(int, file.readline().split()))


class InsertionSort(unittest.TestCase):
    def test1(self):
        arr = load_array("../insert1.txt")
        insertion_sort(arr)
        self.assertEqual(arr, [1, 2, 3, 4, 5])

    def test2(self):
        arr = load_array("../insert2.txt")
        insertion_sort(arr)
        self.assertEqual(arr, [5, 6, 11, 12, 13])

    def test3(self):
        arr = load_array("../insert3.txt")
        insertion_sort(arr)
        self.assertEqual(arr, [-5, -2, -1, 0, 3, 8])


if __name__ == "__main__":
    unittest.main()

File: lab2/insert.cpp
#define CATCH_CONFIG_MAIN
#include "catch.hpp"
#include <fstream>
#include <vector>
using namespace std;
void insertionSort(int arr[], int n) {
  for (int i = 1; i < n; ++i) {
    int key = arr[i];
    int j = i - 1;

    while (j >= 0 && arr[j] > key) {
      arr[j + 1] = arr[j];
      j = j - 1;
    }
    arr[j + 1] = key;
  }
}

vector<int> readFromFile(const string &fileName) {
  ifstream file(fileName);
  vector<int> arr;
  int num;

  if (file.is_open()) {
    while (file >> num) {
      arr.push_back(num);
    }
    file.close();
  } else {
    throw runtime_error("Unable to open file");
  }

  return arr;
}

TEST_CASE("Test", "[insertionSort]") {
  SECTION("test 1") {
    vector<int> arr = readFromFile("../insert1.txt");
    int n = arr.size();
    insertionSort(arr.data(), n);

    vector<int> expected = {1, 2, 3, 4, 5};
    REQUIRE(arr == expected);
  }

  SECTION("test 2") {
    vector<int> arr = readFromFile("../insert2.txt");
    int n = arr.size();
    insertionSort(arr.data(), n);

    vector<int> expected = {5, 6, 11, 12, 13};
    REQUIRE(arr == expected);
  }

  SECTION("test 3") {
    vector<int> arr = readFromFile("../insert3.txt");
    int n = arr.size();
    insertionSort(arr.data(), n);

    vector<int> expected = {-5, -2, -1, 0, 3, 8};
    REQUIRE(arr == expected);
  }
}
